### PR TITLE
Mapcaps

### PR DESCRIPTION
**Pull request by gfiorav**
Still on it

### CODE REVIEW

**Comment by gfiorav** [comment]
Please CR @juanignaciosl !

**Comment by gfiorav** [comment]
Also: we need to discuss need of a Mapcap limit

**Comment by juanignaciosl** [comment]
Can you provide some context about mapcap limit? Do you want me to CR this before?

**Comment by gfiorav** [comment]
Ok hold on, I need to apply the mapcap limit. I'll ping when ready, sorry!

**Comment by juanignaciosl** [comment]
No problem!

**Comment by gfiorav** [comment]
🍏 @juanignaciosl ! Please CR now

File: db/migrate/20160622130822_create_mapcaps.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+Sequel.migration do
+  up do
+    create_table :mapcaps do
+      Uuid :id, primary_key: true, default: 'uuid_generate_v4()'.lit
+      Uuid :visualization_id, type: 'uuid', null: false
+
+      String :export_json, text: true, null: false
```
Comment (juanignaciosl) [inline db/migrate/20160622130822_create_mapcaps.rb]: If content is JSON, use `json` type. `CartoJsonSymbolizerSerializer` and `CartoJsonSymbolizerValidator` will be helpful.

File: app/controllers/carto/api/mapcaps_controller.rb [hunk 2]
```diff
@@ -0,0 +1,75 @@
+# encoding: utf-8
+require_relative '../builder/builder_users_module'
+
+module Carto
+  module Api
+    class MapcapsController < ::Api::ApplicationController
+      include Carto::ControllerHelper
+      include Carto::Builder::BuilderUsersModule
+
+      ssl_required :show, :create, :destroy
```
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcaps_controller.rb]: `index` is probably missing here.

File: app/controllers/carto/api/mapcaps_controller.rb [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+
+      def destroy
+        @mapcap.destroy
+
+        render_jsonp(Carto::Api::MapcapPresenter.new(@mapcap).to_poro)
+      end
+
+      private
+
+      def load_visualization
```
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcaps_controller.rb]: How come there isn't any security check?

File: app/controllers/carto/api/mapcaps_controller.rb [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+        raise Carto::LoadError.new("Visualization not found: #{visualization_id}")
+      end
+
+      def owners_only
+        raise Carto::UnauthorizedError.new unless @visualization.is_writable_by_user(current_user)
+      end
+
+      MAX_MAPCAPS_PER_MAP = 1
+
+      def ensure_only_one_mapcap
```
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcaps_controller.rb]: This belongs to the model.

File: app/controllers/carto/api/mapcaps_controller.rb [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+        previous_mapcaps = @visualization.mapcaps # already ordered from newer to older
+
+        previous_mapcaps[MAX_MAPCAPS_PER_MAP..-1].each(&:destroy) unless previous_mapcaps.count <= MAX_MAPCAPS_PER_MAP
+      end
+
+      def load_mapcap
+        mapcap_id = params[:id]
+
+        @mapcap = Carto::Mapcap.find(mapcap_id)
```
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcaps_controller.rb]: Use `uuid_parameter` method.

File: app/controllers/carto/builder/public/embeds_controller.rb [hunk 6]
```diff
@@ -8,7 +8,7 @@ class EmbedsController < BuilderController
 
         ssl_required :show, :show_protected
 
-        before_filter :load_visualization, only: [:show, :show_protected]
+        before_filter :load_visualization, :load_vizjson, only: [:show, :show_protected]
```
Comment (juanignaciosl) [inline app/controllers/carto/builder/public/embeds_controller.rb]: This makes `load_vizjson` to run twice.

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 2 above)
Comment (gfiorav) [inline app/controllers/carto/api/mapcaps_controller.rb]: oh! Right thanks!

File: app/models/carto/map.rb [hunk 7]
```diff
@@ -64,6 +28,34 @@ class Carto::Map < ActiveRecord::Base
...
+  def data_layers
+    layers.select(&:carto?)
+  end
+
+  def user_layers
+    layers.select(&:user_layer?)
+  end
+
+  def carto_and_torque_layers
+    layers.select(&:carto?) + layers.select(&:torque?)
```
Comment (juanignaciosl) [inline app/models/carto/map.rb]: This changes the order. You probably want `layers.select { |l| l.carto? || l.torque? }`.

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 4 above)
Comment (gfiorav) [inline app/controllers/carto/api/mapcaps_controller.rb]: I was divided on this: the fact that there can only be one mapcap is inherent to it being used for embeds, not to the mapcap nature itself.

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 5 above)
Comment (gfiorav) [inline app/controllers/carto/api/mapcaps_controller.rb]: Interesting, will change!

File: app/models/carto/mapcap.rb [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+    def regenerate_visualization
+      regenerated_visualization = build_visualization_from_json_export(export_json)
+
+      regenerated_visualization.user = regenerated_visualization.map.user = visualization.user
+
+      repopulate_ids(regenerated_visualization)
+    end
+
+    def ids_json
+      JSON.load(self[:ids_json]).with_indifferent_access
```
Comment (juanignaciosl) [inline app/models/carto/mapcap.rb]: Related to migration change, this can be probably implemented with serializers.

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 5 above)
Comment (gfiorav) [inline app/controllers/carto/api/mapcaps_controller.rb]: Would you also add it to `load_visualization` or did you leave it out on purpose?

File: app/controllers/carto/builder/public/embeds_controller.rb (hunk 6 above)
Comment (gfiorav) [inline app/controllers/carto/builder/public/embeds_controller.rb]: ok, I see, I'll put it in another line, thanks!

File: db/migrate/20160622130822_create_mapcaps.rb (hunk 1 above)
Comment (gfiorav) [inline db/migrate/20160622130822_create_mapcaps.rb]: I was wondering about that. I'll take `db/migrate/20160307094600_analysis.rb` as a reference

File: app/models/carto/map.rb (hunk 7 above)
Comment (gfiorav) [inline app/models/carto/map.rb]: oh, torque layers always go on top in named maps templates, my bad, changing!

File: app/models/carto/visualization.rb [hunk 9]
```diff
@@ -348,6 +356,14 @@ def all_users_with_read_permission
     permission.users_with_permissions([CartoDB::Visualization::Member::PERMISSION_READONLY]).push(user)
   end
 
+  def mapcaps
+    Carto::Mapcap.where(visualization_id: id).order('created_at DESC')
```
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: Add the association instead of this.

File: app/models/carto/mapcap.rb (hunk 8 above)
Comment (gfiorav) [inline app/models/carto/mapcap.rb]: absolutely, looking into the serialiser solution!

File: app/models/carto/visualization.rb [hunk 10]
```diff
@@ -348,6 +356,14 @@ def all_users_with_read_permission
...
+  end
+
+  def latest_mapcap
+    Carto::Mapcap.where(visualization_id: id).order('created_at DESC').limit(1).first
```
Comment (juanignaciosl) [inline app/models/carto/visualization.rb]: Add an index at `mapcaps.visualization_id`.

File: app/models/carto/visualization.rb (hunk 9 above)
Comment (gfiorav) [inline app/models/carto/visualization.rb]: I'm finding it difficult to scope the association, that's why I used this.

I'll give it a second though and also add a cascade (forgot about that)

File: app/models/map.rb [hunk 11]
```diff
@@ -79,6 +79,12 @@ def after_save
   end
 
   def notify_map_change
+    visualization = visualizations.first
+
+    force_notify_map_change unless visualization && visualization.latest_mapcap
```
Comment (juanignaciosl) [inline app/models/map.rb]: Could you comment why is this needed?

File: app/models/visualization/member.rb [hunk 12]
```diff
@@ -648,7 +648,7 @@ def invalidate_redis_cache
       end
 
       def save_named_map
-        return if type == TYPE_REMOTE
+        return if type == TYPE_REMOTE || latest_mapcap
```
Comment (juanignaciosl) [inline app/models/visualization/member.rb]: Consider a refactor of this lines that allows you to use `.exists?` to these `latest_mapcap` existance checks.

File: app/services/carto/visualizations_export_service_2.rb [hunk 13]
```diff
@@ -98,12 +98,20 @@ def build_layer_from_hash(exported_layer, order:)
         options[option_key] = nil if options.has_key?(option_key)
       end
 
+      options_with_indifferent_access = options.with_indifferent_access
```
Comment (juanignaciosl) [inline app/services/carto/visualizations_export_service_2.rb]: Makes sense 👍 . I'd maybe add a comment here explaining that this is done in order to keep coherence with Layer serializers.

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcaps_controller.rb]: I don't agree. The decision of "there can only be one mapcap" is determined by embeds, but once that decision is made and there's only one mapcap, it's in its "nature" and it should belong to the model. Otherwise you open the door to having multiple Mapcaps, since other parts of the code can do it.

On the other hand, if we actually want several Mapcaps in some cases, this is the way to go.

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 5 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcaps_controller.rb]: It depends, in some cases we allow both name and id in the same parameter for visualizations. Careful with that.

File: app/models/map.rb (hunk 11 above)
Comment (gfiorav) [inline app/models/map.rb]: we don't want to update the named map if there is a mpacap since it's useless, the same template will be generated over and over. We include a condition to not update if there is a mapcap.

The thing is that unless you do this (or pass an argument) when you want to update the template map once for the mapcap creation, it won't run since there is a mapcap, so you create a separate routine that runs without checking: force_xxx. The name is pretty bad, if you can come up with a better one I'm all ears.

File: app/models/visualization/member.rb (hunk 12 above)
Comment (gfiorav) [inline app/models/visualization/member.rb]: ah, you mean that `latest_mapcap` could be `[]`? Right now it's imposible, it should be null. But, better safe than sorry, I agree.

File: app/services/carto/visualizations_export_service_2.rb (hunk 13 above)
Comment (gfiorav) [inline app/services/carto/visualizations_export_service_2.rb]: Ok, I'll add the comment, thanks!

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 4 above)
Comment (gfiorav) [inline app/controllers/carto/api/mapcaps_controller.rb]: I can't agree :( 

This *will* change very shortly, so maybe we can agree to leave like this as a temporal hack?

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 5 above)
Comment (gfiorav) [inline app/controllers/carto/api/mapcaps_controller.rb]: In this case I'm explicitly treating it like an `id` and thats fine with me. I'll use it over there as well.

File: app/models/visualization/member.rb (hunk 12 above)
Comment (juanignaciosl) [inline app/models/visualization/member.rb]: No, I should've explained this better. if you do `latest_mapcap` just to check that `latest_mapcap` is not nill, you're bringing the full object from database. Nevertheless, if you use `.exists?`, it's just a select  1, probably index-only. 

Analogous example:

```ruby
irb(main):002:0> Carto::Layer.where(id: 'abe26d2a-a308-4e97-bd29-9a5d860f4f74').order('id asc').limit(1).first
  Carto::Layer Load (0.7ms)  SELECT "layers".* FROM "layers" WHERE "layers"."id" = 'abe26d2a-a308-4e97-bd29-9a5d860f4f74' ORDER BY id asc LIMIT 1
=> nil
irb(main):003:0> Carto::Layer.where(id: 'abe26d2a-a308-4e97-bd29-9a5d860f4f74').exists?
  Carto::Layer Exists (0.8ms)  SELECT 1 AS one FROM "layers" WHERE "layers"."id" = 'abe26d2a-a308-4e97-bd29-9a5d860f4f74' LIMIT 1
=> false
```

Since `save_named_map` is a really common operation, small optimizations can be interesting.

Comment (gfiorav) [inline app/models/visualization/member.rb]: ah, very nice explanation! thanks, i'll change it!

File: app/models/map.rb (hunk 11 above)
Comment (juanignaciosl) [inline app/models/map.rb]: No, I guessed that this was not trivial, that's why I asked for a _comment_ in the code :-)

About naming, maybe method extraction creates a wrong abstraction. `notify_map_change` was a "notification" (ey, map, there's a change, do your stuff), so name is ok, but the new method is not. I'd inline the new method within the check:

```ruby
unless visualization && visualization.latest_mapcap
     update_related_named_maps
     invalidate_vizjson_varnish_cache
end
```

File: app/controllers/carto/api/mapcaps_controller.rb (hunk 4 above)
Comment (juanignaciosl) [inline app/controllers/carto/api/mapcaps_controller.rb]: Ok, if this will change there's no need for further discussion :-)

File: app/models/map.rb (hunk 11 above)
Comment (gfiorav) [inline app/models/map.rb]: yeah, but the problem is that you want to do it once when you create the mapcap, but you can't call this method since it will check for the mapcap and do nothing :D

Comment (juanignaciosl) [inline app/models/map.rb]: No, no. The previous code is just a suggested refactor for `notify_map_change`. I think that line 84 should be changed by my suggestion, as `force_notify_map_change` is only called there and it makes the code less readable.

Comment (gfiorav) [inline app/models/map.rb]: Ah, I see. Duplicate code for readability? I agree in this case

File: app/models/carto/visualization.rb (hunk 10 above)
Comment (gfiorav) [inline app/models/carto/visualization.rb]: Adding 👍

File: app/models/map.rb (hunk 11 above)
Comment (juanignaciosl) [inline app/models/map.rb]: No, not duplicated. I'd remove `force_notify_map_change` method, which is no longer used.

Comment (gfiorav) [inline app/models/map.rb]: Ok: major mistake on my side.

Mapcap is supposed to call `force_notify_map_change` instead of `notify_map_change`

Sorry for that!

File: db/migrate/20160622130822_create_mapcaps.rb (hunk 1 above)
Comment (gfiorav) [inline db/migrate/20160622130822_create_mapcaps.rb]: I won't use it for `export_json`since it's used as a string but i'll do it for `ids_json`

File: app/models/map.rb (hunk 11 above)
Comment (juanignaciosl) [inline app/models/map.rb]: Everything makes sense then :-) I'd suggest `propagate_map_changes` as a better name for `force_notify_map_change ` then, but I haven't an strong opinion, up to you.

File: db/migrate/20160622130822_create_mapcaps.rb (hunk 1 above)
Comment (juanignaciosl) [inline db/migrate/20160622130822_create_mapcaps.rb]: Well, since it _is_ a json database also grants data consistency, I'd also use json for it.

**Comment by gfiorav** [comment]
Thanks @juanignaciosl , took your comments and made some profound changes.

Let's have @javitonino for the second round!

**Comment by juanignaciosl** [comment]
Great, I'll be around for next comments :-)

File: db/migrate/20160628105307_create_mapcaps.rb [hunk 14]
```diff
@@ -0,0 +1,25 @@
+Sequel.migration do
+  up do
+    create_table :mapcaps do
+      Uuid :id, primary_key: true, default: 'uuid_generate_v4()'.lit
+      Uuid :visualization_id, type: 'uuid', null: false
```
Comment (javitonino) [inline db/migrate/20160628105307_create_mapcaps.rb]: No foreign key?

File: lib/carto/named_maps/template.rb [hunk 15]
```diff
@@ -20,7 +20,7 @@ class Template
       DATAVIEW_TEMPLATE_OPTIONS = [:column, :aggregation, :aggregationColumn, :aggregation_column, :operation].freeze
 
       def initialize(visualization)
-        # TODO: Remove when it's safe to assume thais confussion wont' happen.
+        # TODO: Remove when it's safe to assume thais confussion won't happen.
```
Comment (javitonino) [inline lib/carto/named_maps/template.rb]: Selective type fixing :P (thais)

File: db/migrate/20160628105307_create_mapcaps.rb [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+
+      String :export_json, null: false, type: 'json'
+      String :ids_json, null: false, type: 'json'
+
+      DateTime :created_at, default: Sequel::CURRENT_TIMESTAMP
+    end
+
+    alter_table :mapcaps do
+      add_index [:visualization_id, :created_at]
```
Comment (javitonino) [inline db/migrate/20160628105307_create_mapcaps.rb]: For current behaviour, with only one mapcap per visualization, this index is overkill. Will be useful if we allow multiple mapcaps per viz, but we don't know that for certain. Also, there has been some talk of adding channels to this, where the logical index would be `[viz_id, channel, created_at]`. I guess what I'm trying to say is that this seems overengineered, and I would leave only the viz_id in the index. Adding indices is a relatively harmless operation (specially using `CONCURRENTLY`) so I would add them as needed.

Comment (juanignaciosl) [inline db/migrate/20160628105307_create_mapcaps.rb]: Fair enough, when I suggested the index I didn't realize that mapcaps where unique.

File: app/controllers/carto/api/mapcaps_controller.rb [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+      end
+
+      def show
+        render_jsonp(Carto::Api::MapcapPresenter.new(@mapcap).to_poro)
+      end
+
+      def destroy
+        @mapcap.destroy
+
+        render_jsonp(Carto::Api::MapcapPresenter.new(@mapcap).to_poro)
```
Comment (javitonino) [inline app/controllers/carto/api/mapcaps_controller.rb]: Usually destroy returns 204 and empty body. Any reason to go against convention?

Comment (gfiorav) [inline app/controllers/carto/api/mapcaps_controller.rb]: not at all, adding!

File: db/migrate/20160628105307_create_mapcaps.rb (hunk 16 above)
Comment (gfiorav) [inline db/migrate/20160628105307_create_mapcaps.rb]: removing!

File: app/models/carto/map.rb [hunk 18]
```diff
@@ -64,6 +28,34 @@ class Carto::Map < ActiveRecord::Base
...
+  def torque_layers
+    layers.select(&:torque?)
+  end
+
+  def other_layers
+    layers.reject(&:carto?)
+          .reject(&:tiled?)
+          .reject(&:background?)
+          .reject(&:gmapsbase?)
+          .reject(&:wms?)
```
Comment (javitonino) [inline app/models/carto/map.rb]: This seems to be equivalent to `layers.reject(&:named_map_layer?)`, isn't it?

File: app/models/carto/visualization.rb [hunk 19]
```diff
@@ -348,6 +357,14 @@ def all_users_with_read_permission
     permission.users_with_permissions([CartoDB::Visualization::Member::PERMISSION_READONLY]).push(user)
   end
 
+  def mapcaped?
+    mapcaps.exists?
+  end
+
+  def latest_mapcap
+    mapcaps.limit(1).first
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: Is limit required here?

File: lib/carto/named_maps/template.rb (hunk 15 above)
Comment (gfiorav) [inline lib/carto/named_maps/template.rb]: haha changing

File: app/models/visualization/member.rb [hunk 20]
```diff
@@ -680,6 +680,18 @@ def map
         @map ||= ::Map.where(id: map_id).first
       end
 
+      def mapcaps
+        Carto::Mapcap.latest_for_visualization(id)
+      end
+
+      def latest_mapcap
+        mapcaps.limit(1).first
```
Comment (javitonino) [inline app/models/visualization/member.rb]: Same limit comment here.

File: lib/carto/named_maps/template.rb [hunk 21]
```diff
@@ -57,9 +57,9 @@ def name
       def placeholders
         placeholders = {}
 
-        layers = @visualization.map.layers
+        layers = @visualization.layers
 
-        last_index, carto_layers_visibility_placeholders = layer_visibility_placeholders(layers.select(&:carto_layer?))
+        last_index, carto_layers_visibility_placeholders = layer_visibility_placeholders(layers.select(&:carto?))
```
Comment (javitonino) [inline lib/carto/named_maps/template.rb]: Just saw this a bit below, this could be rewritten as `@visualization.data_layers` and `@visualization.torque_layers` but it's pretty readable either way, so up to you.

File: app/models/carto/visualization.rb [hunk 22]
```diff
@@ -348,6 +357,14 @@ def all_users_with_read_permission
     permission.users_with_permissions([CartoDB::Visualization::Member::PERMISSION_READONLY]).push(user)
   end
 
+  def mapcaped?
```
Comment (javitonino) [inline app/models/carto/visualization.rb]: Don't know this english rule by heart, but this feels like the last p should be double: `def mapcapped?`

:trollface:

File: app/models/carto/mapcap.rb [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+    end
+
+    def self.latest_for_visualization(visualization_id)
+      where(visualization_id: visualization_id).order('created_at DESC')
+    end
+
+    private
+
+    def generate_export_json
+      self.export_json = export_visualization_json_hash(visualization_id, visualization.user)
```
Comment (javitonino) [inline app/models/carto/mapcap.rb]: This checks that the visualization is derived and raises in other cases. Can you check what happens when you call the mapcaps API with a canonical visualization. Most likely caught by `rescue_from StandardError` but as this raises an string, maybe not.

File: app/models/carto/mapcap.rb [hunk 24]
```diff
@@ -0,0 +1,79 @@
...
+
+    belongs_to :visualization, class_name: Carto::Visualization, foreign_key: 'visualization_id'
+
+    serialize :ids_json, ::Carto::CartoJsonSerializer
+    serialize :export_json, ::Carto::CartoJsonSerializer
+
+    after_save :notify_map_change, :update_named_map
+    after_destroy :notify_map_change
+
+    before_validation :generate_export_json, :generate_ids_json
```
Comment (javitonino) [inline app/models/carto/mapcap.rb]: Cool trick!

**Comment by javitonino** [comment]
Looking great. I left some comments, I'll take another look after them.

File: app/models/carto/visualization.rb (hunk 19 above)
Comment (gfiorav) [inline app/models/carto/visualization.rb]: Yes! I wan't it to only fetch 1, no need for the rest

File: app/models/carto/map.rb (hunk 18 above)
Comment (gfiorav) [inline app/models/carto/map.rb]: yeah, but I wan't to stick the the `other` part.

File: app/models/carto/mapcap.rb (hunk 24 above)
Comment (gfiorav) [inline app/models/carto/mapcap.rb]: 😎

File: app/models/carto/mapcap.rb (hunk 23 above)
Comment (gfiorav) [inline app/models/carto/mapcap.rb]: You're right about that case, hadn't though about it. I'll add a filter to control it. Any suggestions for return codes in that case?

File: app/models/carto/visualization.rb (hunk 22 above)
Comment (gfiorav) [inline app/models/carto/visualization.rb]: u lie

File: db/migrate/20160628105307_create_mapcaps.rb (hunk 14 above)
Comment (gfiorav) [inline db/migrate/20160628105307_create_mapcaps.rb]: added!

File: lib/carto/named_maps/template.rb (hunk 21 above)
Comment (gfiorav) [inline lib/carto/named_maps/template.rb]: yeah, good catch. I'll leave it as is for the moment, I trust this to be more future-proof

**Comment by gfiorav** [comment]
hey @javitonino I made some changes. Review when you can please!

Thanks!

File: db/migrate/20160628105307_create_mapcaps.rb [hunk 25]
```diff
@@ -0,0 +1,18 @@
+Sequel.migration do
+  up do
+    create_table :mapcaps do
+      foreign_key :visualization_id, :visualizations, type: 'uuid', null: false
```
Comment (javitonino) [inline db/migrate/20160628105307_create_mapcaps.rb]: Please check that this creates an index. PostgreSQL doesn't create an index by default for foreign keys.

File: app/models/carto/visualization.rb (hunk 19 above)
Comment (javitonino) [inline app/models/carto/visualization.rb]: `.first` does just that, with the advantage of using AR caching if the relation has been preloaded:
```
irb(main):010:0> m.layers.first
  Carto::Layer Load (1.7ms)  SELECT "layers".* FROM "layers" INNER JOIN "layers_maps" ON "layers"."id" = "layers_maps"."layer_id" WHERE "layers_maps"."map_id" = 'b52448b6-03b2-4040-b715-ecafd86a25ec' ORDER BY "order" LIMIT 1
[...]
irb(main):011:0> m.layers
  Carto::Layer Load (0.8ms)  SELECT "layers".* FROM "layers" INNER JOIN "layers_maps" ON "layers"."id" = "layers_maps"."layer_id" WHERE "layers_maps"."map_id" = 'b52448b6-03b2-4040-b715-ecafd86a25ec' ORDER BY "order"
[...]
irb(main):012:0> m.layers.first
<no query>
```

File: app/models/carto/map.rb (hunk 18 above)
Comment (javitonino) [inline app/models/carto/map.rb]: I see the point :sparkles:

File: app/models/carto/mapcap.rb (hunk 23 above)
Comment (javitonino) [inline app/models/carto/mapcap.rb]: 422? It's what we use everywhere for semantic errors. Or maybe just 404 and change the message to "Derived visualization not found".

File: spec/models/carto/mapcap_spec.rb [hunk 26]
```diff
@@ -0,0 +1,175 @@
...
+require 'spec_helper_min'
+require 'support/helpers'
+
+describe Carto::Mapcap do
+  include Carto::Factories::Visualizations
+
+  before(:all) do
+    @user = FactoryGirl.create(:carto_user, private_tables_enabled: true)
+
+    @map, _, _, @visualization = create_full_visualization(@user)
```
Comment (javitonino) [inline spec/models/carto/mapcap_spec.rb]: I'd get the other parameters (table and table_visualization) so you can properly destroy them later with `destroy_full_visualization`.

File: spec/requests/carto/api/mapcaps_controller_spec.rb [hunk 27]
```diff
@@ -0,0 +1,193 @@
...
+    dummy
+  end
+
+  before(:all) do
+    FactoryGirl.create(:carto_feature_flag, name: 'editor-3', restricted: false)
+
+    @user = FactoryGirl.create(:carto_user)
+    @intruder = FactoryGirl.create(:carto_user)
+
+    _, _, _, @visualization = create_full_visualization(@user)
```
Comment (javitonino) [inline spec/requests/carto/api/mapcaps_controller_spec.rb]: Same here.

File: lib/carto/named_maps/template.rb (hunk 21 above)
Comment (javitonino) [inline lib/carto/named_maps/template.rb]: :palm_tree:

**Comment by javitonino** [comment]
Left some more minor comments, act on them as you like. Check the index is being created, and this looks like a nice :package:.

File: db/migrate/20160628105307_create_mapcaps.rb (hunk 25 above)
Comment (gfiorav) [inline db/migrate/20160628105307_create_mapcaps.rb]: Nice catch!

**Comment by gfiorav** [comment]
@javitonino I've reviewed them, thanks! I'm going for a 1 man smoke test session and then dropping the bomb.

**Comment by gfiorav** [comment]
Hey @javitonino I've made some changes, can you review?

Basically, I wasn't adding the permission on regeneration and that prevented Mapcaps from being created on private organization maps (named map needs a permission for auth tokens when org + private viz).

I've kind of been selective about what I load. The principle is "add as we need". We can add as we go, it won't affect old mapcaps (it has nothing to do with persisted stuff).

**Comment by gfiorav** [comment]
UPDATE: I also refactored the limit.

File: app/models/carto/mapcap.rb [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+
+    validates :ids_json, carto_json_symbolizer: true
+    validates :export_json, carto_json_symbolizer: true
+
+    def regenerate_visualization
+      regenerated_visualization = build_visualization_from_hash_export(export_json)
+
+      regenerated_visualization.user = regenerated_visualization.map.user = visualization.user
+      regenerated_visualization.permission = visualization.permission
```
Comment (javitonino) [inline app/models/carto/mapcap.rb]: Hmm, thinking about this... Is the mapcap regenerated when touching the visualization privacy/permissions? Does this work?:
 1. Create visualization, set to public
 2. Mapcap it
 3. Make visualization private
 4. Load embed

It should have auth tokens (in orgs or protected) but that implies updating the named map. Have you tested this flow?

**Comment by javitonino** [comment]
The permission stuff inspired a major concern: I think privacy changes to the visualization are not reflected in the mapcap right now. Maybe we should update the mapcap + namedmap when changing visualization privacy/permissions, but that sounds kinda hard to do without changing anything else. :/

File: app/models/carto/mapcap.rb (hunk 28 above)
Comment (gfiorav) [inline app/models/carto/mapcap.rb]: I see your concern. Token are used in two situations:

+ Org + private
+ Password protected

In both cases the embed controller will forbid you from accessing a
visualization your now allowed to see.

File: app/controllers/carto/api/mapcaps_controller.rb [hunk 29]
```diff
@@ -0,0 +1,78 @@
...
+      end
+
+      private
+
+      def load_visualization
+        visualization_id = uuid_parameter(:visualization_id)
+
+        @visualization = Carto::Visualization.find(visualization_id)
+
+        raise ActiveRecord::RecordNotFound if @visualization.canonical?
```
Comment (javitonino) [inline app/controllers/carto/api/mapcaps_controller.rb]: Bonus comments. I think this is better as `unless visualization.derived?`. Current code would let slide and remote visualizations, which also do not work.

Comment (gfiorav) [inline app/controllers/carto/api/mapcaps_controller.rb]: Yup, I'll change it if need to commit somethings else :P

File: app/models/carto/mapcap.rb (hunk 28 above)
Comment (javitonino) [inline app/models/carto/mapcap.rb]: Uhm, and the other way around. The visualization is initially set to org + private, so it requires auth tokens. Then it changes to public. Does the named map require auth tokens still, and does the embed send them?

Maybe I'm overthinking this, if you are confident, go for it. Anyway, my concerns are only related to mapcaps themselves and does not affects all code, so we can always fix it later if we find some corner case.

**Comment by gfiorav** [comment]
About permission changes: broken for auth -> public, easily fixed publishing a new mapcap. Will take care of it later.

Please check the latest changes @javitonino :

+ On delete cascade in the migration
+ `save_named_map` returns true if mapcap (fixes privacy revert)

**Comment by javitonino** [comment]
:gift_heart:
